### DISCUSSION TITLE
How to insert animations

### DISCUSSION THREAD

**Discussion by cristianadrielbraun** [AllenDang/giu#447] [Q&A] [answered]
What would be the right way to aproach this?
I mean simple animation, gif-styled.
Like loading dots, or stuff like that
I tried to actually load a gif but without success.
Thanks in advance :D

**Comment by gucio321** [comment]
I'd do as follows:
- load gif with golang api
- split gif into frames
- create imgui textur for each of them
- save textures somewhere
- display animation e.g.
```golang
// on top
const deltaTime = 20
var currentFrame int = 0
var frames []giu.Texture // lets say that we've all animation frames in here

// inside of loop func
giu.Image(frames[currentFrame]),

// somewhere else, out of loop func (e.g. in main)
go func() {
   for range time.Tick(deltaTime) {
      currentFrame++
      if currentFrame == len(frames) {
         currentFrame = 0
      }
   }
}()
```

**Reply by gucio321** [reply]
well, you can also use gif's delay time instead of hard-coded delta-time param
reffering to golang's GIF api:
```golang
type GIF struct {
	Image []*[image](https://pkg.go.dev/image).[Paletted](https://pkg.go.dev/image#Paletted) // The successive images.
	Delay [][int](https://pkg.go.dev/builtin#int)
...
}
```

**Comment by gucio321** [comment, marked as answer]
@yomono 4u
```golang
package main

import (
	"bytes"
	"image/gif"
	"log"
	"os"
	"time"

	"github.com/AllenDang/giu"
)

const gifFilepath = "./golang.gif"

var (
	frames       []*giu.Texture
	gifImg       *gif.GIF
	currentFrame int
)

func loop() {
	// load textures
	if frames[0] == nil {
		for i, frame := range gifImg.Image {
			// lol, this is the most magic thing in go i ever seen :D
			i := i
			giu.NewTextureFromRgba(giu.ImageToRgba(frame), func(t *giu.Texture) {
				frames[i] = t
			})
		}
	}

	giu.SingleWindow().Layout(
		giu.Image(frames[currentFrame]),
	)
}

func main() {
	wnd := giu.NewMasterWindow("GIF renderer [example]", 640, 480, 0)

	gifFileData, err := os.ReadFile(gifFilepath)
	if err != nil {
		log.Fatal(err)
	}

	gifImg, err = gif.DecodeAll(bytes.NewReader(gifFileData))
	if err != nil {
		log.Fatal(err)
	}

	frames = make([]*giu.Texture, len(gifImg.Image))

	go func() {
		for {
			time.Sleep(time.Duration(gifImg.Delay[currentFrame]*10) * time.Millisecond)
			giu.Update()
			currentFrame++
			if currentFrame == len(frames) {
				currentFrame = 0
			}
		}
	}()

	wnd.Run(loop)
}
```

@AllenDang what do you think about putting this code in examples? it may be interessting

**Comment by cristianadrielbraun** [comment]
Sir, you are amazing. 
I asked for copper and you gave me gold.
That code worked flawesly without any adjustments:

https://user-images.githubusercontent.com/31445936/152565704-9711ed4a-a1e5-4894-8964-3651b5b586bc.mp4

And yes @AllenDang, it would be a nice example tu add!

Thank you @gucio321 you are a master


PS: Can you explain to me the `i := i ` on the code? I don't quite get what is going on there

**Reply by gucio321** [reply]
> And yes @AllenDang, it would be a nice example tu add!

yah, I created a PR with this example #448 

> PS: Can you explain to me the i := i on the code? I don't quite get what is going on there

lol, just like I said: `// lol, this is the most magic thing in go i ever seen :D` but without it, the `i` value inside of callback function is always last iterated index.
I really don't know what is going on here :D

**Reply by gucio321** [reply]
oh, one more @yomono you can also use `go:embed` to embed your gif and not worry about the image is in the same folder you're running the app. I've just added it to #448

**Reply by cristianadrielbraun** [reply]
Yes, actually I'm embedding all my resources also thanks to YOU, since you helped me with that too here #399
So I really have a lot to thank you for 😄

**Reply by gucio321** [reply]
no, problem ;-)

@AllenDang  in case of this code I've a few notes:
it is a bit strange that I need to use NewTexture... niside of loop func, so
- panic whn using it outside of loop should come from giu (not mainthread)
- we can add NewTextureFromRgbaInAdvance and create some load queue

**Comment by cristianadrielbraun** [comment]
I didn't do any more tests yet, but I guess you should also be using channels or WaitGroups to control the infinite loop inside the goroutine.
So you can start/stop the gifImg.Delay when needed. 
Because, you know, an infinite loop is never a good idea
